traktorml.py

```diff
@@ -0,0 +1,196 @@
+        self.deselected_columns = []
+        self.deselected_run_ids = []
+        self.current_sort = None
+        self.filter_selected_values = {}
+        self.filter_deselected_values = defaultdict(list)
+
+
+def set_data_table_to_experiment_view(experiment_view: ExperimentRunsView, data_table: DataTable):
+    """Helper function to set the values of a Textual data table to an experiment view."""
+    data_table.clear(columns=True)
+    for col in experiment_view.selected_columns:
+        data_table.add_column(label=col, key=col)
+    for i, row in enumerate(experiment_view.selected_data):
+        data_table.add_row(*row, key=row[0])
+
+
+class TraktorML(App):
+
+    TITLE = "Traktor ML"
+    CSS = "DataTable {height: 1fr}"
+    BINDINGS = [
+        ("s", "sort", "Sort by column"),
+        ("p", "clear_selections", "Clear selections"),
+        ("f", "filter_on_values", "Filter on values"),
+        ("d", "deselect_values", "Deselect values"),
+        ("c", "remove_columns", "Remove columns"),
+        ("r", "remove_row", "Remove Row"),
+        ("q", "quit", "Quit"),
+    ]
+
+    experiments = get_experiments()
+    selected_experiment: Union[Experiment, None] = None
+    experiment_views: Dict[str, ExperimentRunsView] = {}
+
+    @property
+    def experiment_view(self):
+        if self.selected_experiment.name not in self.experiment_views:
+            self.experiment_views[self.selected_experiment.name] = ExperimentRunsView(
+                experiment_runs_df=get_experiment_runs_df(self.selected_experiment)
+            )
+        return self.experiment_views[self.selected_experiment.name]
+
+    def compose(self) -> ComposeResult:
+        yield Header()
+        yield Footer()
+        yield Select(
+            options=[(experiment.name, experiment.id) for experiment in self.experiments],
+            prompt="Select experiment"
+        )
+        yield ScrollableContainer(DataTable(), id="run_table")
+
+    def on_key(self, event: Key) -> None:
+        if event.key == "s":
+            run_table = self.query_one(DataTable)
+            cursor_coordinate = run_table.cursor_coordinate
+            row_key, col_key = run_table.coordinate_to_cell_key(cursor_coordinate)
+
+            # Filter out missing values to allow sorting
+            self.experiment_view.filter_deselected_values[col_key.value].append(MISSING_VALUE)
+            set_data_table_to_experiment_view(self.experiment_view, run_table)
+
+            if self.experiment_view.current_sort is None or self.experiment_view.current_sort[0] != col_key:
+                is_reverse = False
+                run_table.sort(col_key, reverse=is_reverse)
+            else:
+                is_reverse = not self.experiment_view.current_sort[1]
+                run_table.sort(col_key, reverse=is_reverse)
+
+            self.experiment_view.current_sort = (col_key, is_reverse)
+            run_table.cursor_coordinate = cursor_coordinate
+
+        if event.key == "c":
```

Comment:
seems to be a lot of **almost** repetition in these `if event.key == ...` blocks. Can it be generalized.

Reply:
You’re right, there is still a lot of repetition, even if I moved some of it into the set_data_table_to_experiment_view function. I think we should wait with more generalization until we experiment a bit more with textual.